File: hard/1510-stone-game-iv/solution.java
/**
 * 1510. Stone Game IV
 * Time: O(n * sqrt(n))
 * Space: O(n)
 */
class Solution {
    public boolean winnerSquareGame(int n) {
        // win[i] = true if the player to move with i stones wins
        boolean[] win = new boolean[n + 1];
        for (int i = 1; i <= n; i++) {
            for (int k = 1; k * k <= i; k++) {
                if (!win[i - k * k]) {
                    win[i] = true; // leave the opponent in a losing position
                    break;
                }
            }
        }
        return win[n];
    }
}

File: hard/1510-stone-game-iv/solution.py
class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        """
        1510. Stone Game IV
        Time: O(n * sqrt(n))
        Space: O(n)
        """
        # win[i] = True if the player to move with i stones wins
        win = [False] * (n + 1)
        for i in range(1, n + 1):
            k = 1
            while k * k <= i:
                if not win[i - k * k]:
                    win[i] = True  # leave the opponent in a losing position
                    break
                k += 1
        return win[n]
